File: src/util/units/Bitrate.java
package util.units;

import util.parsing.ParsesFromString;
import util.parsing.ParsesToString;
import util.parsing.StringParseStrategy;
import util.parsing.StringParseStrategy.From;

import static java.lang.Integer.parseInt;
import static util.parsing.StringParseStrategy.To.TO_STRING_METHOD;

/**
 * Simple class for media bit rate. Internally represents the value as int.
 * Unit is kbps.
 *
 * @author Martin Polakovic
 */
@StringParseStrategy(
    from = From.ANNOTATED_METHOD,
    to = TO_STRING_METHOD,
    exFrom = { IndexOutOfBoundsException.class, NumberFormatException.class, IllegalArgumentException.class }
)
public class Bitrate implements Comparable<Bitrate> {
    private static final String UNIT = "kbps";
    private final int bitrate;

    /**
     * @param value bit rate value in kb per second. Use -1 if not available.
     */
    public Bitrate(int value){
        if(value<-1) throw new IllegalArgumentException("Bitrate value must be -1 or larger");
        bitrate = value;
    }

    /***
     * @param s number as string with optional unit 'kbps' appended. White spaces
     * are removed.
     */
    @ParsesFromString
    public Bitrate(String s){
        this(val(s));
    }

    /** @return bit rate value in kb per second. */
    public int getValue() {
        return bitrate;
    }

    @Override
    public int compareTo(Bitrate o) {
        return Integer.compare(bitrate, o.bitrate);
    }

    /**
     * Appends ' kbps' string after value. If no value available it returns "".
     * For example: "320 kbps" or "N/A"
     * @return string representation of the object
     */
    @Override
    @ParsesToString
    public String toString() {
        return bitrate == -1 ? "" : bitrate + " " + UNIT;
    }

    /** @return true if the value is the same. */
    @Override
    public boolean equals(Object o) {
        return (this == o) || (o instanceof Bitrate && ((Bitrate) o).bitrate == bitrate);
    }

    @Override
    public int hashCode() {
        return 97 * 7 + this.bitrate;
    }




    private static int val(String s) throws IndexOutOfBoundsException, NumberFormatException {
        if(s.endsWith(UNIT)) s=s.substring(0, s.length()-UNIT.length());
        s = s.trim();
        return s.isEmpty() ? -1 : parseInt(s);
    }
}

File: src/sp/it/pl/util/units/Bitrate.kt
package sp.it.pl.util.units

import sp.it.pl.util.dev.Dependency
import sp.it.pl.util.functional.Try

/** Media bit rate with kb/s unit. Supports variable and unknown value. */
data class Bitrate(/** Value in kb/s, 0 if unknown, and negative if variable. */ val value: Int): Comparable<Bitrate> {

    /** @return true iff the value is unknown */
    fun isUnknown() = value==VALUE_NA

    /** @return true iff this represents variable bitrate */
    fun isVariable() = value<0

    /** @return true iff this represents constant bitrate */
    fun isConstant() = !isVariable()

    override fun compareTo(other: Bitrate) = Integer.compare(value, other.value)

    /** For example: "~320 kbps" or "N/A". */
    @Dependency("fromString")
    override fun toString() = when {
        value==VALUE_NA -> VALUE_S_NA
        value<0 -> "$VALUE_S_VARIABLE${-value} $UNIT"
        else -> "$value $UNIT"
    }

    companion object {
        private val VALUE_NA = 0
        private val VALUE_S_NA = "n/a"
        private val VALUE_S_VARIABLE = "~"
        private val UNIT = "kbps"

        @Dependency("toString")
        @JvmStatic
        fun fromString(s: String): Try<Bitrate> {
            if (VALUE_S_NA==s) return Try.ok(Bitrate(VALUE_NA))

            return try {
                var v = s
                v = v.trim()
                if (v.startsWith(VALUE_S_VARIABLE)) v = v.substringAfter(VALUE_S_VARIABLE)
                if (v.endsWith(UNIT)) v = v.substring(0, v.length-UNIT.length)
                v = v.trim()

                Try.ok(if (v.isEmpty()) Bitrate(VALUE_NA) else Bitrate(v.toInt()))
            } catch (e: IndexOutOfBoundsException) {
                Try.error(e)
            } catch (e: NumberFormatException) {
                Try.error(e)
            } catch (e: IllegalArgumentException) {
                Try.error(e)
            }
        }
    }

}